File: verify_project.py
import os
import sys
import json

# Force stdout to UTF-8 for Windows console support
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')

print("====================================================")
print(" PLACEMENTOS AUTOMATED VERIFICATION & TEST SUITE ")
print("====================================================\n")

base_dir = os.path.dirname(os.path.abspath(__file__))

total_tests = 0
passed_tests = 0
failed_tests = 0

def assert_test(condition, message):
    global total_tests, passed_tests, failed_tests
    total_tests += 1
    if condition:
        passed_tests += 1
        print(f"  [PASS] {message}")
    else:
        failed_tests += 1
        print(f"  [FAIL] {message}")

# 1. FILE EXISTENCE & INTEGRITY
print("1. Checking Core Project Files & Directory Structure...")

required_files = [
    'index.html',
    'css/main.css',
    'css/components.css',
    'css/pages.css',
    'js/app.js',
    'js/models/StateStore.js',
    'js/models/StorageService.js',
    'js/models/ReadinessEngine.js',
    'js/models/RecommendationEngine.js',
    'js/models/Validators.js',
    'js/components/UIComponents.js',
    'js/components/OnboardingWizard.js',
    'js/components/ChartEngine.js',
    'js/services/ExportService.js',
    'js/views/DashboardView.js',
    'js/views/ProfileView.js',
    'js/views/CareerGoalsView.js',
    'js/views/RoadmapView.js',
    'js/views/TaskManagerView.js',
    'js/views/SkillTrackerView.js',
    'js/views/PlacementReadinessView.js',
    'js/views/AnalyticsView.js',
    'js/views/HistoryView.js',
    'js/views/AchievementsView.js',
    'js/views/ExportView.js',
    'js/views/SettingsView.js',
    'presentation.html',
    'PlacementOS_Presentation.pptx'
]

for rel_path in required_files:
    full_path = os.path.join(base_dir, rel_path)
    assert_test(os.path.exists(full_path), f"File exists: {rel_path}")

# 2. HTML STRUCTURE & NAVIGATION CONTAINERS
print("\n2. Checking HTML Structure & Navigation Containers...")
index_path = os.path.join(base_dir, 'index.html')
with open(index_path, 'r', encoding='utf-8') as f:
    html_content = f.read()

assert_test('<meta name="viewport"' in html_content, "index.html has responsive viewport meta tag")
assert_test('<div id="app">' in html_content, "index.html has #app main container")
assert_test('type="module" src="js/app.js"' in html_content, "index.html imports js/app.js as ES6 module")

routes = [
    'dashboard', 'profile', 'career-goals', 'roadmap',
    'task-manager', 'skill-tracker', 'placement-readiness',
    'analytics', 'history', 'achievements', 'export', 'settings'
]

for r in routes:
    page_id = f'id="page-{r}"'
    assert_test(page_id in html_content, f"HTML contains view container `#page-{r}`")

# 3. ES6 MODULE IMPORTS & EXPORTS SANITY
print("\n3. Verifying ES6 Imports & View Class Definitions...")

app_js_path = os.path.join(base_dir, 'js', 'app.js')
with open(app_js_path, 'r', encoding='utf-8') as f:
    app_js_content = f.read()

view_classes = [
    'DashboardView', 'ProfileView', 'CareerGoalsView', 'RoadmapView',
    'TaskManagerView', 'SkillTrackerView', 'PlacementReadinessView',
    'AnalyticsView', 'HistoryView', 'AchievementsView', 'ExportView', 'SettingsView'
]

for cls in view_classes:
    assert_test(f'import {{ {cls} }}' in app_js_content, f"app.js imports {cls}")
    view_file_path = os.path.join(base_dir, 'js', 'views', f'{cls}.js')
    if os.path.exists(view_file_path):
        with open(view_file_path, 'r', encoding='utf-8') as vf:
            v_content = vf.read()
            assert_test(f'export class {cls}' in v_content, f"{cls}.js defines and exports class `{cls}`")
            assert_test('static render(' in v_content, f"{cls} implements `static render(container, appState)`")

# 4. CSS DESIGN TOKENS & RESPONSIVE GRID
print("\n4. Checking CSS Styling System & Responsive Utilities...")
css_comp_path = os.path.join(base_dir, 'css', 'components.css')
with open(css_comp_path, 'r', encoding='utf-8') as f:
    css_content = f.read()

assert_test('.grid-cols-4' in css_content, "components.css contains `.grid-cols-4` grid layout")
assert_test('.notification-badge' in css_content, "components.css defines `.notification-badge` styling")
assert_test('.modal-overlay' in css_content, "components.css defines `.modal-overlay` component")

print("\n====================================================")
print(f" FINAL RESULTS: {passed_tests}/{total_tests} VERIFICATION CHECKS PASSED")
if failed_tests == 0:
    print(" PROJECT STATUS: 100% HEALTHY & FULLY VERIFIED!")
else:
    print(f" {failed_tests} ISSUES ENCOUNTERED.")
print("====================================================\n")

File: verify_project.js
/* ==========================================================================
   verify_project.js - Automated Verification & Test Suite for PlacementOS
   ========================================================================== */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

console.log('====================================================');
console.log(' 🚀 PLACEMENTOS AUTOMATED VERIFICATION & TEST SUITE ');
console.log('====================================================\n');

let totalTests = 0;
let passedTests = 0;
let failedTests = 0;

function assert(condition, message) {
  totalTests++;
  if (condition) {
    passedTests++;
    console.log(`  ✅ [PASS] ${message}`);
  } else {
    failedTests++;
    console.error(`  ❌ [FAIL] ${message}`);
  }
}

// 1. FILE EXISTENCE & INTEGRITY CHECKS
console.log('🔍 1. Checking Core Project Files & Structure...');

const requiredFiles = [
  'index.html',
  'css/index.css',
  'css/components.css',
  'css/pages.css',
  'js/app.js',
  'js/models/StateStore.js',
  'js/models/StorageService.js',
  'js/models/ReadinessEngine.js',
  'js/models/RecommendationEngine.js',
  'js/models/Validators.js',
  'js/components/UIComponents.js',
  'js/components/OnboardingWizard.js',
  'js/components/ChartEngine.js',
  'js/services/ExportService.js',
  'js/views/DashboardView.js',
  'js/views/ProfileView.js',
  'js/views/CareerGoalsView.js',
  'js/views/RoadmapView.js',
  'js/views/TaskManagerView.js',
  'js/views/SkillTrackerView.js',
  'js/views/PlacementReadinessView.js',
  'js/views/AnalyticsView.js',
  'js/views/HistoryView.js',
  'js/views/AchievementsView.js',
  'js/views/ExportView.js',
  'js/views/SettingsView.js'
];

requiredFiles.forEach(file => {
  const fullPath = path.join(__dirname, file);
  assert(fs.existsSync(fullPath), `File exists: ${file}`);
});

// 2. IMPORT & DATA ENGINE TESTS
console.log('\n🧠 2. Testing Data Engines & Business Logic...');

async function runLogicTests() {
  try {
    const { ReadinessEngine } = await import('./js/models/ReadinessEngine.js');
    const { RecommendationEngine } = await import('./js/models/RecommendationEngine.js');
    const { Validators } = await import('./js/models/Validators.js');
    const { StorageService } = await import('./js/models/StorageService.js');

    // Test Readiness Engine Calculation
    const mockTasks = [
      { id: '1', category: 'DSA', status: 'completed' },
      { id: '2', category: 'Core CS', status: 'completed' },
      { id: '3', category: 'Web Dev', status: 'pending' },
    ];
    const mockSkills = [
      { id: '1', level: 'Advanced' },
      { id: '2', level: 'Intermediate' }
    ];
    const mockUser = { name: 'Test Student', cgpa: 8.5, targetRole: 'SDE-1' };

    const scoreData = ReadinessEngine.calculateScore(mockTasks, mockSkills, mockUser);
    assert(typeof scoreData.overall === 'number', 'Readiness score returns a valid number');
    assert(scoreData.overall >= 0 && scoreData.overall <= 100, `Score is within 0-100 range (Got: ${scoreData.overall})`);
    assert(scoreData.breakdown !== undefined, 'Score breakdown metrics generated');

    // Test Recommendation Engine
    const recs = RecommendationEngine.getRecommendations({ tasks: mockTasks, skills: mockSkills, user: mockUser });
    assert(Array.isArray(recs), 'Recommendations engine returns an array');
    assert(recs.length > 0, `Generated ${recs.length} actionable recommendations`);

    // Test Validators
    assert(Validators.isEmail('student@college.edu') === true, 'Email validator accepts valid email');
    assert(Validators.isEmail('invalid-email') === false, 'Email validator rejects invalid email');
    assert(Validators.isValidCGPA(8.7) === true, 'CGPA validator accepts 8.7');
    assert(Validators.isValidCGPA(11.5) === false, 'CGPA validator rejects 11.5');
    assert(Validators.isValidCGPA(-1.0) === false, 'CGPA validator rejects -1.0');

    // Test Storage Schema Validation
    const validState = StorageService.getInitialState();
    assert(StorageService.validateBackupJSON(validState) === true, 'Initial state satisfies backup schema validation');
    assert(StorageService.validateBackupJSON({ foo: 'bar' }) === false, 'Backup validator rejects malformed state');

  } catch (err) {
    console.error('Logic test error:', err);
    assert(false, `Logic test execution error: ${err.message}`);
  }
}

// 3. VIEW EXPORT TESTS
console.log('\n🖼️ 3. Validating View Renderers & Component Exports...');

async function runViewTests() {
  const views = [
    { name: 'DashboardView', file: './js/views/DashboardView.js' },
    { name: 'ProfileView', file: './js/views/ProfileView.js' },
    { name: 'CareerGoalsView', file: './js/views/CareerGoalsView.js' },
    { name: 'RoadmapView', file: './js/views/RoadmapView.js' },
    { name: 'TaskManagerView', file: './js/views/TaskManagerView.js' },
    { name: 'SkillTrackerView', file: './js/views/SkillTrackerView.js' },
    { name: 'PlacementReadinessView', file: './js/views/PlacementReadinessView.js' },
    { name: 'AnalyticsView', file: './js/views/AnalyticsView.js' },
    { name: 'HistoryView', file: './js/views/HistoryView.js' },
    { name: 'AchievementsView', file: './js/views/AchievementsView.js' },
    { name: 'ExportView', file: './js/views/ExportView.js' },
    { name: 'SettingsView', file: './js/views/SettingsView.js' }
  ];

  for (const v of views) {
    try {
      const mod = await import(v.file);
      assert(mod[v.name] !== undefined, `Module exports class: ${v.name}`);
      assert(typeof mod[v.name].render === 'function', `Class has static render method: ${v.name}`);
    } catch (err) {
      assert(false, `Failed to load view module ${v.name}: ${err.message}`);
    }
  }
}

// RUN ALL TESTS
(async () => {
  await runLogicTests();
  await runViewTests();

  console.log('\n====================================================');
  console.log(` 📊 VERIFICATION RESULTS: ${passedTests}/${totalTests} TESTS PASSED`);
  if (failedTests === 0) {
    console.log(' ✨ ALL VERIFICATION CHECKS PASSED PERFECTLY!');
  } else {
    console.log(` ⚠️ ${failedTests} CHECKS FAILED.`);
  }
  console.log('====================================================\n');
})();
